From cc6ed40245dd594c3caa9e46960dec8d95b3f917 Mon Sep 17 00:00:00 2001
From: Tomas Halman <thalman@redhat.com>
Date: Thu, 13 Mar 2025 17:37:51 +0100
Subject: [PATCH] Enumerate object with escaped characters in name
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

This patch fixes enumeration when DN in LDAP server
contains special characters.

The libldb expects that '\' is followed by two hex digits
in filter. Strings like '\#' must be sanitized into '\5c#'
before they are used for searching.

Resolves: https://github.com/SSSD/sssd/issues/7876

Reviewed-by: Alejandro López <allopez@redhat.com>
Reviewed-by: Dan Lavu <dlavu@redhat.com>
(cherry picked from commit 158b4cdb7ac62fde1280f50a5d678f80d0e99015)

Reviewed-by: Alejandro López <allopez@redhat.com>
(cherry picked from commit 116d6221c8dba014a12b7ca93eff62fd3f0f314f)
Reviewed-by: Tomáš Halman <thalman@redhat.com>
---
 src/db/sysdb_search.c | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/src/db/sysdb_search.c b/src/db/sysdb_search.c
index 49362beb0..b45df2358 100644
--- a/src/db/sysdb_search.c
+++ b/src/db/sysdb_search.c
@@ -784,6 +784,7 @@ static errno_t sysdb_enum_dn_filter(TALLOC_CTX *mem_ctx,
 {
     TALLOC_CTX *tmp_ctx = NULL;
     char *dn_filter;
+    char *sanitized_dn;
     const char *fqname;
     errno_t ret;
 
@@ -814,11 +815,18 @@ static errno_t sysdb_enum_dn_filter(TALLOC_CTX *mem_ctx,
     }
 
     for (size_t i = 0; i < ts_res->count; i++) {
+        ret = sss_filter_sanitize_dn(tmp_ctx,
+                                     ldb_dn_get_linearized(ts_res->msgs[i]->dn),
+                                     &sanitized_dn);
+        if (ret != EOK) {
+            goto done;
+        }
         dn_filter = talloc_asprintf_append(
                                   dn_filter,
                                   "(%s=%s)",
                                   SYSDB_DN,
-                                  ldb_dn_get_linearized(ts_res->msgs[i]->dn));
+                                  sanitized_dn);
+        talloc_free(sanitized_dn);
         if (dn_filter == NULL) {
             ret = ENOMEM;
             goto done;
-- 
2.52.0

